### PR TITLE
docs: Add documentation on how to add a new package.

### PR DESCRIPTION
**Pull request by eiffel-fl**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

Add documentation on how to add new package to CBL Mariner.

###### Change Log  

- Add documentation

###### Does this affect the toolchain?  

**NO**

### CODE REVIEW

**Review by mfrw — COMMENTED** [review]
Can we please retarget to `main` if we want it for `2.0` or `3.0-dev` if we want it for `3.0` :)

File: toolkit/docs/building/add-package.md [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+```
+
+## Building the package
+
+We have all the files needed to build the package, let's build it:
+
+```bash
+$ cd toolkit
+# First, we need to prepare to get the toolchain:
+$ ./scripts/setuplkgtoolchain.sh
```
Comment (christopherco) [inline toolkit/docs/building/add-package.md:296]: This script is currently only available in the 3.0 tree. Let's target this PR towards 3.0-dev. Eventually, the need for this script will go away when we publish our 3.0 packages to our package server, similar to 2.0, where the toolkit will natively pull packages from our public package repo.

File: toolkit/docs/building/add-package.md [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+root@azure-linux [ ~ ]# ig --help
+Collection of gadgets for containers
+
+Usage:
+  ig [command]
+...
+^Croot@azure-linux [ ~ ]# ig trace exec --host > /tmp/ig.log &
+[1] 706
+root@azure-linux [ ~ ]# ls /dev/null
+ig-0.25.1-1.x86_64.rpm
```
Comment (christopherco) [inline toolkit/docs/building/add-package.md:426]:
```suggestion
ig-0.25.0-1.azl3.x86_64.rpm
```
I think the filename should be this?

File: toolkit/docs/building/add-package.md [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+$ sftp -P 10022 root@localhost                                                                                                                                          francis/marinade-ig *% u=
+Welcome to CBL-Mariner 2.0.20240223 (x86_64)
+(root@localhost) Password:
+Connected to localhost.
+sftp> put ../out/RPMS/x86_64/ig-0.25.0-1.azl3.x86_64.rpm
+Uploading ../out/RPMS/x86_64/ig-0.25.0-1.azl3.x86_64.rpm to /root/ig-0.25.0-1.azl3.x86_64.rpm
+ig-0.25.0-1.azl3.x86_64.rpm                                                                                                                                                                                          100%   16MB   8.0MB/s   00:02
+sftp> bye
+# Go back to the VM and install the package:
+root@azure-linux [ ~ ]# tdnf install -y ig-0.25.1-1.x86_64.rpm
```
Comment (christopherco) [inline toolkit/docs/building/add-package.md:411]:
```suggestion
root@azure-linux [ ~ ]# tdnf install -y ig-0.25.0-1.azl3.x86_64.rpm
```
I think the filename should be this?

File: toolkit/docs/building/add-package.md [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+```
+
+We can now write our signature file:
+
+```bash
+$ sha256sum *.tar.gz
+ba1c9153ee67cd9d1ebd31021ca3848c1099979c229b26bea6ad05470ac486ac  ig-0.25.0-govendor-v1.tar.gz
+3d283e0f6e758b36e291e483e79e7a08c4ce2965a6c11f5c1ab671ced7015a6e  v0.25.0.tar.gz
+# Edit the ig.signatures.json.
+$ $EDITOR ig.signatures.json
```
Comment (christopherco) [inline toolkit/docs/building/add-package.md:277]: Just wanted to mention, the toolkit does have an option to compute the hashes on your behalf, which could save time & automate the `*.signatures.json` edits - https://github.com/microsoft/azurelinux/blob/3.0-dev/toolkit/docs/building/building.md#source-hashes

**Review by christopherco — CHANGES_REQUESTED** [review]
@eiffel-fl this is a nice walkthrough on creating a new package for Mariner / Azure Linux, especially with the golang vendoring quirk! Really like that you used a specific example - `ig` - to do the walkthrough.

My primary request is to retarget this PR to `3.0-dev` since some of the steps are specific to building on `3.0-dev`. Rest of the comments are minor.

**Review by christopherco — APPROVED** [review]

File: toolkit/docs/building/add-package.md [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+$ ./usr/bin/ig --help
+Collection of gadgets for containers
+
+Usage:
+  ig [command]
+...
+```
+
+Congratulations! You successfully built a CBL-Mariner package!
+
```
Comment (dmcilvaney) [inline toolkit/docs/building/add-package.md:316]: We have a few automatic tools in the toolkit to help here, setting up a full VM can be a pain.

Feel free to edit the formatting, the suggestion box didn't like using backticks.
```suggestion
## Testing in a Container

The toolkit can generate an azl containers for you, and auto launch them:
```bash
# Start the test container
sudo make containerized-rpmbuild MODE=test SRPM_PACK_LIST='ig' DAILY_BUILD_ID=3-0-2024028
...
~/repos/temp/CBL-Mariner_TEMP3/toolkit/scripts/containerized-build/create_container_build.sh 
...
Populating Intermediate SRPMs...
...
Setting up tools...
Setting up mounts...
Importing chroot into docker...
Chroot is up-to-date
Checking if build env is up-to-date...
...
 -----------------------------------------------------------------------------------------
 ----------------------------------- MARINER BUILDER ! ----------------------------------- 
 -----------------------------------------------------------------------------------------
******************************************************************************************
* To see this menu again, run  show_help
*
* Some tips:
*     Use local RPMs to satify dependencies:        run enable_local_repo
...

# Create a rpm repo based on the contents of ./out/RPMS/
enable_local_repo
...
-------- enabling local repo ---------
...
# Install the new package
tdnf install /mnt/RPMS/x86_64/ig-0.25.0-1.azl3.x86_64.rpm
ig --help
```

**Review by dmcilvaney — COMMENTED** [review]
Thanks for the guide! I had a few nits/suggestions.

File: toolkit/docs/building/add-package.md [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+Each package relies on 2 different files:
+
+1. A SPEC file: it contains information on the software and how to build and package it.
+2. A signature file: this JSON file contains hashes of `tar.gz` archives used to get the sources.
+
+We will now take a look at each of these files.
+
+### The SPEC file
+
+`CBL Mariner` uses RPM packages, so you will need to write a [`SPEC` file](https://rpm-software-management.github.io/rpm/manual/spec.html)
```
Comment (dmcilvaney) [inline toolkit/docs/building/add-package.md:23]:
```suggestion
`Azure Linux` uses RPM packages, so you will need to write a [`SPEC` file](https://rpm-software-management.github.io/rpm/manual/spec.html)
```

File: toolkit/docs/building/add-package.md [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+
+# Files which will be part of the resulting RPM package.
+%files
+%license LICENSE
+%license LICENSE-bpf.txt
+%{_bindir}/ig
+
+%changelog
+* Tue Mar 14 2023 Francis Laniel <flaniel@linux.microsoft.com> - 0.25.0-1
+- Original version for CBL-Mariner
```
Comment (dmcilvaney) [inline toolkit/docs/building/add-package.md:98]:
```suggestion
- Original version for Azure Linux
```

File: toolkit/docs/building/add-package.md [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+149181 blocs
+$ ./usr/bin/ig --help
+Collection of gadgets for containers
+
+Usage:
+  ig [command]
+...
+```
+
+Congratulations! You successfully built a CBL-Mariner package!
```
Comment (dmcilvaney) [inline toolkit/docs/building/add-package.md:315]:
```suggestion
Congratulations! You successfully built an Azure Linux package!
```

**Comment by christopherco** [comment]
> Hi!
> 
> Thank you for the approval! Before we merge this, do you know where we should link this file? Indeed, this commit only adds a new file without linking it to an existing part of the documentation.
> 
> Best regards.

Linking to this file from the main toolkit readme under its own subheading seems like the best spot - https://github.com/microsoft/azurelinux/blob/3.0-dev/toolkit/README.md

**Pull request merged**